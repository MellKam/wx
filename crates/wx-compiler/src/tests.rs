use super::*;
use indoc::indoc;
use insta::{assert_binary_snapshot, assert_yaml_snapshot};
use string_interner::{StringInterner, backend::StringBackend};

#[allow(unused)]
struct TestCase {
    files: files::Files,
    interner: StringInterner<StringBackend>,
    ast: ast::parser::ParserResult,
    hir: hir::builder::BuilderResult,
    mir: mir::MIR,
    module: wasm::Module,
    bytecode: Vec<u8>,
}

impl TestCase {
    fn assert_snapshot(name: &str, source: &str) {
        let mut interner = StringInterner::new();
        let mut files = files::Files::new();
        let file_id = files
            .add(name.to_string() + ".wx", source.to_string())
            .unwrap();

        let ast =
            ast::parser::Parser::parse(file_id, &files.get(file_id).unwrap().source, &mut interner);
        assert_yaml_snapshot!(name.to_string() + "_ast", ast);

        if ast.diagnostics.len() > 0 {
            return;
        }

        let hir = hir::Builder::build(&ast.ast, &mut interner);
        assert_yaml_snapshot!(name.to_string() + "_hir", hir);

        let mir = mir::Builder::build(&hir.hir);
        assert_yaml_snapshot!(name.to_string() + "_mir", mir);

        let module = wasm::Builder::build(&mir).unwrap();
        assert_yaml_snapshot!(name.to_string() + "_wasm", module);

        let bytecode = wasm::Encoder::encode(&module, &interner);
        assert_binary_snapshot!((name.to_string() + ".wasm").as_ref(), bytecode.clone());
    }
}

#[test]
fn factorial() {
    TestCase::assert_snapshot(
        "factorial",
        indoc! {"
        export func factorial(n: i32): i32 {
            if n <= 1 { return 1 };
            n * factorial(n - 1)
        }
    "},
    );
}

#[test]
fn pow() {
    TestCase::assert_snapshot(
        "pow",
        indoc! {"
        export func pow(base: i32, mut exp: i32): i32 {
            if exp < 0 {
                return 0; 
            };
            
            local mut result: i32 = 1;
            loop {
                if exp == 0 {
                break result; 
                };
                
                result *= base;
                exp -= 1;
            }
        }
    "},
    );
}

#[test]
fn func_pointers() {
    TestCase::assert_snapshot(
        "func_pointers",
        indoc! {"
        func add(a: i32, b: i32): i32 { a + b }
        func sub(a: i32, b: i32): i32 { a - b }

        func apply(
            binop: func(i32, i32) -> i32, 
            a: i32, 
            b: i32,
        ): i32 {
            binop(a, b)
        }

        export func main(): i32 {
            local a = apply(add, 5, 10);
            local b = apply(sub, 10, 5);
            local c: i32 = 5;

            a + b
        }
    "},
    );
}
