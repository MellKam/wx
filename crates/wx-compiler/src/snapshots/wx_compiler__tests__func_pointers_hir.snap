---
source: crates/wx-compiler/src/tests.rs
expression: hir
---
hir:
  file_id: 0
  functions:
    - name:
        symbol: 1
        span:
          start: 5
          end: 8
      ty:
        params:
          - Primitive: I32
          - Primitive: I32
        result:
          Primitive: I32
      stack:
        scopes:
          - kind: Block
            label: ~
            parent: ~
            locals:
              - name:
                  symbol: 2
                  span:
                    start: 9
                    end: 10
                ty:
                  Primitive: I32
                mutability: ~
                accesses:
                  - span:
                      start: 32
                      end: 33
                    kind: Read
              - name:
                  symbol: 4
                  span:
                    start: 17
                    end: 18
                ty:
                  Primitive: I32
                mutability: ~
                accesses:
                  - span:
                      start: 36
                      end: 37
                    kind: Read
            inferred_type:
              Primitive: I32
            expected_type:
              Primitive: I32
      block:
        kind:
          Block:
            scope_index: 0
            expressions: []
            result:
              kind:
                Binary:
                  operator:
                    kind: Add
                    span:
                      start: 34
                      end: 35
                  left:
                    kind:
                      Local:
                        scope_index: 0
                        local_index: 0
                    span:
                      start: 32
                      end: 33
                    ty:
                      Primitive: I32
                  right:
                    kind:
                      Local:
                        scope_index: 0
                        local_index: 1
                    span:
                      start: 36
                      end: 37
                    ty:
                      Primitive: I32
              span:
                start: 32
                end: 37
              ty:
                Primitive: I32
        span:
          start: 30
          end: 39
        ty:
          Primitive: I32
    - name:
        symbol: 5
        span:
          start: 45
          end: 48
      ty:
        params:
          - Primitive: I32
          - Primitive: I32
        result:
          Primitive: I32
      stack:
        scopes:
          - kind: Block
            label: ~
            parent: ~
            locals:
              - name:
                  symbol: 2
                  span:
                    start: 49
                    end: 50
                ty:
                  Primitive: I32
                mutability: ~
                accesses:
                  - span:
                      start: 72
                      end: 73
                    kind: Read
              - name:
                  symbol: 4
                  span:
                    start: 57
                    end: 58
                ty:
                  Primitive: I32
                mutability: ~
                accesses:
                  - span:
                      start: 76
                      end: 77
                    kind: Read
            inferred_type:
              Primitive: I32
            expected_type:
              Primitive: I32
      block:
        kind:
          Block:
            scope_index: 0
            expressions: []
            result:
              kind:
                Binary:
                  operator:
                    kind: Sub
                    span:
                      start: 74
                      end: 75
                  left:
                    kind:
                      Local:
                        scope_index: 0
                        local_index: 0
                    span:
                      start: 72
                      end: 73
                    ty:
                      Primitive: I32
                  right:
                    kind:
                      Local:
                        scope_index: 0
                        local_index: 1
                    span:
                      start: 76
                      end: 77
                    ty:
                      Primitive: I32
              span:
                start: 72
                end: 77
              ty:
                Primitive: I32
        span:
          start: 70
          end: 79
        ty:
          Primitive: I32
    - name:
        symbol: 6
        span:
          start: 86
          end: 91
      ty:
        params:
          - Function: 0
          - Primitive: I32
          - Primitive: I32
        result:
          Primitive: I32
      stack:
        scopes:
          - kind: Block
            label: ~
            parent: ~
            locals:
              - name:
                  symbol: 7
                  span:
                    start: 97
                    end: 102
                ty:
                  Function: 0
                mutability: ~
                accesses:
                  - span:
                      start: 166
                      end: 171
                    kind: Read
              - name:
                  symbol: 2
                  span:
                    start: 132
                    end: 133
                ty:
                  Primitive: I32
                mutability: ~
                accesses:
                  - span:
                      start: 172
                      end: 173
                    kind: Read
              - name:
                  symbol: 4
                  span:
                    start: 145
                    end: 146
                ty:
                  Primitive: I32
                mutability: ~
                accesses:
                  - span:
                      start: 175
                      end: 176
                    kind: Read
            inferred_type:
              Primitive: I32
            expected_type:
              Primitive: I32
      block:
        kind:
          Block:
            scope_index: 0
            expressions: []
            result:
              kind:
                Call:
                  callee:
                    kind:
                      Local:
                        scope_index: 0
                        local_index: 0
                    span:
                      start: 166
                      end: 171
                    ty:
                      Function: 0
                  arguments:
                    - kind:
                        Local:
                          scope_index: 0
                          local_index: 1
                      span:
                        start: 172
                        end: 173
                      ty:
                        Primitive: I32
                    - kind:
                        Local:
                          scope_index: 0
                          local_index: 2
                      span:
                        start: 175
                        end: 176
                      ty:
                        Primitive: I32
              span:
                start: 166
                end: 177
              ty:
                Primitive: I32
        span:
          start: 160
          end: 179
        ty:
          Primitive: I32
    - name:
        symbol: 8
        span:
          start: 193
          end: 197
      ty:
        params: []
        result:
          Primitive: I32
      stack:
        scopes:
          - kind: Block
            label: ~
            parent: ~
            locals:
              - name:
                  symbol: 2
                  span:
                    start: 217
                    end: 218
                ty:
                  Primitive: I32
                mutability: ~
                accesses:
                  - span:
                      start: 300
                      end: 301
                    kind: Read
              - name:
                  symbol: 4
                  span:
                    start: 250
                    end: 251
                ty:
                  Primitive: I32
                mutability: ~
                accesses:
                  - span:
                      start: 304
                      end: 305
                    kind: Read
              - name:
                  symbol: 9
                  span:
                    start: 283
                    end: 284
                ty:
                  Primitive: I32
                mutability: ~
                accesses: []
            inferred_type:
              Primitive: I32
            expected_type:
              Primitive: I32
      block:
        kind:
          Block:
            scope_index: 0
            expressions:
              - kind:
                  LocalDeclaration:
                    name:
                      symbol: 2
                      span:
                        start: 217
                        end: 218
                    scope_index: 0
                    local_index: 0
                    expr:
                      kind:
                        Call:
                          callee:
                            kind:
                              Function: 2
                            span:
                              start: 221
                              end: 226
                            ty:
                              Function: 1
                          arguments:
                            - kind:
                                Function: 0
                              span:
                                start: 227
                                end: 230
                              ty:
                                Function: 0
                            - kind:
                                Int: 5
                              span:
                                start: 232
                                end: 233
                              ty:
                                Primitive: I32
                            - kind:
                                Int: 10
                              span:
                                start: 235
                                end: 237
                              ty:
                                Primitive: I32
                      span:
                        start: 221
                        end: 238
                      ty:
                        Primitive: I32
                span:
                  start: 211
                  end: 238
                ty: Unit
              - kind:
                  LocalDeclaration:
                    name:
                      symbol: 4
                      span:
                        start: 250
                        end: 251
                    scope_index: 0
                    local_index: 1
                    expr:
                      kind:
                        Call:
                          callee:
                            kind:
                              Function: 2
                            span:
                              start: 254
                              end: 259
                            ty:
                              Function: 1
                          arguments:
                            - kind:
                                Function: 1
                              span:
                                start: 260
                                end: 263
                              ty:
                                Function: 0
                            - kind:
                                Int: 10
                              span:
                                start: 265
                                end: 267
                              ty:
                                Primitive: I32
                            - kind:
                                Int: 5
                              span:
                                start: 269
                                end: 270
                              ty:
                                Primitive: I32
                      span:
                        start: 254
                        end: 271
                      ty:
                        Primitive: I32
                span:
                  start: 244
                  end: 271
                ty: Unit
              - kind:
                  LocalDeclaration:
                    name:
                      symbol: 9
                      span:
                        start: 283
                        end: 284
                    scope_index: 0
                    local_index: 2
                    expr:
                      kind:
                        Int: 5
                      span:
                        start: 292
                        end: 293
                      ty:
                        Primitive: I32
                span:
                  start: 277
                  end: 293
                ty: Unit
            result:
              kind:
                Binary:
                  operator:
                    kind: Add
                    span:
                      start: 302
                      end: 303
                  left:
                    kind:
                      Local:
                        scope_index: 0
                        local_index: 0
                    span:
                      start: 300
                      end: 301
                    ty:
                      Primitive: I32
                  right:
                    kind:
                      Local:
                        scope_index: 0
                        local_index: 1
                    span:
                      start: 304
                      end: 305
                    ty:
                      Primitive: I32
              span:
                start: 300
                end: 305
              ty:
                Primitive: I32
        span:
          start: 205
          end: 307
        ty:
          Primitive: I32
  enums: []
  globals: []
  exports:
    - Function:
        func_index: 3
diagnostics:
  - severity: Warning
    code: E2009
    message: unused variable
    labels:
      - style: Primary
        file_id: 0
        range:
          start: 283
          end: 284
        message: ""
    notes: []
